crates/handlers/src/upstream_oauth2/mod.rs

```diff
@@ -124,10 +142,32 @@ fn client_credentials_for_provider(
         },
 
         UpstreamOAuthProviderTokenAuthMethod::SignInWithApple => {
-            let params = client_secret.ok_or(ProviderCredentialsError::MissingClientSecret)?;
-            let params: SignInWithApple = serde_json::from_str(&params)?;
+            let client_secret =
+                client_secret.ok_or(ProviderCredentialsError::MissingClientSecret)?;
+
+            let params: SignInWithApple = serde_json::from_str(&client_secret)
+                .map_err(|inner| ProviderCredentialsError::InvalidClientSecretJson { inner })?;
+
+            if params.private_key.is_none() && params.private_key_file.is_none() {
+                return Err(ProviderCredentialsError::MissingPrivateKey);
+            }
+
+            if params.private_key.is_some() && params.private_key_file.is_some() {
+                return Err(ProviderCredentialsError::DuplicatePrivateKey);
+            }
 
-            let key = elliptic_curve::SecretKey::from_pkcs8_pem(&params.private_key)?;
+            let private_key_pem = if let Some(private_key) = params.private_key {
+                private_key
+            } else if let Some(private_key_file) = params.private_key_file {
+                tokio::fs::read_to_string(private_key_file)
+                    .await
+                    .map_err(|_| ProviderCredentialsError::MissingPrivateKey)?
```

Comment:
Sorry I missed that, but this should *not* be done when reading from the database. In MAS, the providers are stored in the database, and the config file is just there to sync on startup.

So:

 - don't touch this part, we should store in the database the key already read
 - read the file when we sync the config file with the database, probably somewhere around here:
https://github.com/element-hq/matrix-authentication-service/blob/a3202a690ad677dacbed264c877e5e6831986012/crates/cli/src/sync.rs#L192-L196